### ISSUE TITLE
dataloader which stratifies train-validation split by adata.obs columns

### ISSUE THREAD

**Issue by vitkl** [scverse/scvi-tools#1063]
Would be great to have train validations split balanced by cell labels or batch - which is useful when there are rare populations. 

This can be addressed by adding a 'stratify' option for dataloader and dataslitter like in `sklearn.model_selection.train_test_split(..., stratify=x)`.

**Comment by adamgayoso** [comment]
What do you think @galenxing @romain-lopez ? The data splitters for training can have an `stratify_by_obs_key: str` argument.

**Comment by adamgayoso** [comment]
@watiss we can also implement internally as an option if it works

**Comment by vitkl** [comment]
So, do you think it is possible to add this feature?

**Comment by vitkl** [comment]
Would be great to have this option as

```python
model = SCVI(...)
model.train(..., stratify_validation_split_by_obs_key='key')
```

**Comment by martinkim0** [comment]
@vitkl We could definitely add this feature. Going off of Adam's comment previously, I think it makes sense to name it `stratify_by_obs_key` with an additional argument `stratify_min_size`, which should either raise an error or warning if the minimum split size cannot be achieved.

**Comment by canergen** [comment]
To me it sounds most flexible to be able to pass an obs column with train_indices, valid_indices and test_indices and leave it up to the user/developer how to define those. I think other solutions will limit us to certain use cases. @vitkl Would it capture your use case?

**Comment by vitkl** [comment]
Thanks for following up! An obs column with custom indices would indeed be great. You mean a column with categories named ['train_indices', 'valid_indices', 'test_indices'], right?

**Comment by ori-kron-wis** [comment]
Hi @vitkl 

The idea is to have the data splitter now accepting also a list of np.arrays of indexes, corresponding to train/valid/test indexes, like this: [train_idx,valid_idx,test_idx] + a flag on whether to use them. 
Although there are some validations for the correctness of the passed indexes, the user is responsible for supporting them in the DataSplitter (as they can originate from many logic and scenarios). 

e.g, consider passing stratified train/valid/test indexes based on some column
`adata = scvi.data.synthetic_iid()`

**Comment by vitkl** [comment]
I need some time to understand what you are proposing properly. I am looking for a solution that would work with the default DataSplitter and scVI .train() - rather than implementing a new model to use it.

**Comment by canergen** [comment]
Sorry for the confusion. The suggested change is still an open PR: https://github.com/scverse/scvi-tools/pull/2902. After merging, the command will be to call scVI.train(external_indexing=[np.array(train_indices), np.array(valid_indices), np.array(test_indices)]. The splitter will then use these splits instead of random splits. To yield stratified splitting, https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html will do the job.
